### PR TITLE
Retry on Crypto-Input forwarding

### PR DESCRIPTION
**Pull request by TonySemikin**
#### Pre-Release

- [x] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [x] Check for linter errors (in PR)
- [x] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes

- [x] Fill in empty outTxId values in all older transactions (dev & prod DBs) @davidleomay 
#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

**Comment by TonySemikin** [comment]
We could go through "Test basic user operations" together on Tuesday if you don't mind. Since we worked on scheduled operations, no direct user input should be affected, but still. Functionality is tested via CLI mostly.

File: src/payment/models/crypto-input/crypto-input.service.ts [hunk 1]
```diff
@@ -253,24 +259,59 @@ export class CryptoInputService {
     });
   }
 
-  private async saveAndForward(input: CryptoInput): Promise<void> {
+  private async testCompositeSwaps(
```
Comment (davidleomay) [inline src/payment/models/crypto-input/crypto-input.service.ts:262]: I would rename this method to something like `getReferenceAmounts()`. USDT, USDC and BTC are reference assets. These are the ones we (virtually) exchange on Kraken.

File: src/payment/models/crypto-input/crypto-input.service.ts [hunk 2]
```diff
@@ -253,24 +259,59 @@ export class CryptoInputService {
...
+
+    console.log(`Forwarding inputs (${inputs.length})`);
+
+    for (const input of inputs) {
+      try {
+        if (input.route.type === RouteType.STAKING) {
+          await this.cryptoStakingService.create(input);
+        }
+
+        // forward (only await UTXO)
```
Comment (davidleomay) [inline src/payment/models/crypto-input/crypto-input.service.ts:305]: `(only await UTXO)` is an old comment, can be removed :-)

File: src/payment/models/crypto-input/crypto-input.service.ts [hunk 3]
```diff
@@ -253,24 +259,59 @@ export class CryptoInputService {
...
+      relations: ['route'],
+    });
+
+    console.log(`Forwarding inputs (${inputs.length})`);
+
+    for (const input of inputs) {
+      try {
+        if (input.route.type === RouteType.STAKING) {
+          await this.cryptoStakingService.create(input);
+        }
```
[the remark this answers is not in the document]
Comment (TonySemikin) [reply src/payment/models/crypto-input/crypto-input.service.ts:303]: yeah, makes perfect sense

File: src/payment/models/crypto-input/crypto-input.service.ts (hunk 1 above)
Comment (TonySemikin) [reply src/payment/models/crypto-input/crypto-input.service.ts]: done

File: src/payment/models/crypto-input/crypto-input.service.ts (hunk 2 above)
Comment (TonySemikin) [reply src/payment/models/crypto-input/crypto-input.service.ts]: done

File: src/payment/models/crypto-input/crypto-input.service.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (TonySemikin) [reply src/payment/models/crypto-input/crypto-input.service.ts]: done, but worth short discussion

**Review by davidleomay — APPROVED** [review]

**Pull request merged**